src/common/logging.rs

```diff
@@ -0,0 +1,272 @@
+    match create_dir_if_not_exist(dest_dir_path) {
+        Ok(_) => (),
+        Err(_) => reboot(),
+    }
+
+    let source_tmp_log = Path::new(source_tmp_log_path);
+    if source_tmp_log.exists() && source_tmp_log.is_file() {
+        let file_name = source_tmp_log
+            .file_name()
+            .map(|name| name.to_string_lossy().to_string())
+            .expect("Failed to extract file name from path");
+
+        match copy(
+            PathBuf::from(source_tmp_log),
+            path_append(dest_dir_path, format!("/{}", file_name)),
+        ) {
+            Ok(_) => info!(
+                "Copied {} from {} to {} on data partition",
+                file_name,
+                &source_tmp_log.display(),
+                &dest_dir_path
+            ),
+            Err(why) => error!(
+                "Could not copy {} from {} to {}: {:?}",
+                file_name,
+                &source_tmp_log.display(),
+                &dest_dir_path,
+                why
+            ),
+        }
+    } else {
+        info!(
+            "File {} does not exist or is not a regular file.",
+            source_tmp_log_path
+        );
+    }
+}
+
+// Helper function to dump logs
+// Caters for state prior to calling pivot_root
+pub fn stage2_init_pre_pivot_root_tmpfs_log_handler() {
+    let stage1_logfile = get_stage_tmpfs_logfile_path(Stage::S1);
+    let stage2_init_logfile = get_stage_tmpfs_logfile_path(Stage::S2Init);
+
+    // copy files to data partition
+    copy_tmpfs_log_to_data_partition(stage1_logfile.as_str(), LOG_PRE_UNMOUNT_DATA_PART_DEST);
+    copy_tmpfs_log_to_data_partition(stage2_init_logfile.as_str(), LOG_PRE_UNMOUNT_DATA_PART_DEST);
+}
+
+// Error handling if an error occurs in stage2-init prior to calling pivot_root
+pub fn stage2_init_pre_pivot_root_err_handler(log_to_balenaos: bool) -> ! {
+    if log_to_balenaos {
+        stage2_init_pre_pivot_root_tmpfs_log_handler();
+    }
+
+    reboot();
+}
+
+// Helper function to dump logs
+// Caters for state after calling pivot_root
+pub fn stage2_init_post_pivot_root_tmpfs_log_handler() {
+    // if pivot_root called, stage1 log will be relative to old root mountpoint
+    let stage1_logfile = format!(
+        "{}/{}",
+        OLD_ROOT_MP,
+        get_stage_tmpfs_logfile_path(Stage::S1)
+    );
+
+    let stage2_init_logfile = get_stage_tmpfs_logfile_path(Stage::S2Init);
+
+    // data part will be relative to old mount point
+    let destination = format!("{}/{}", OLD_ROOT_MP, LOG_PRE_UNMOUNT_DATA_PART_DEST);
+
+    // copy files to data partition
+    copy_tmpfs_log_to_data_partition(stage1_logfile.as_str(), destination.as_str());
+    copy_tmpfs_log_to_data_partition(stage2_init_logfile.as_str(), destination.as_str());
+}
+
+// Error handling if an error occurs in stage2-init after calling pivot_root
+pub fn stage2_init_post_pivot_root_err_handler() -> ! {
+    stage2_init_post_pivot_root_tmpfs_log_handler();
```

Comment:
Does this call need to in a `log_to_balanos` conditional like `stage2_pre_unmount_err_handler()`? If not, why not?

Reply:
good catch